CIP-script-ref-plutus-v1/README.md

```diff
@@ -0,0 +1,118 @@
+
+A simple, backwards compatible mechanism for plutus v1 protocols to satisfy the script witness requirement, without changing the script context and causing breaking changes for Plutus v1 scripts, would alleviate quite literally millions of dollars worth of storage requirements, user pain, and developer frustration.
+
+## Specification
+
+Currently, the relevant parts of the transaction body CDDL are produced below:
+
+```
+transaction =
+  [ transaction_body
+  , transaction_witness_set
+  ...
+  ]
+
+transaction_body =
+  { 0 : set<transaction_input>             ; inputs
+  ...
+  , ? 7 : auxiliary_data_hash
+  ...
+  , ? 11 : script_data_hash
+  ...
+  , ? 13 : nonempty_set<transaction_input> ; collateral inputs
+  ...
+  , ? 18 : nonempty_set<transaction_input> ; reference inputs
+  }
+
+post_alonzo_transaction_output =
+  {
+  ...
+  , ? 3 : script_ref   ; script reference
+  }
+
+transaction_witness_set =
+  {
+  , ? 3: [* plutus_v1_script ]
+  ...
+  , ? 6: [* plutus_v2_script ]
+  , ? 7: [* plutus_v3_script ]
+  }
+```
+
+In order to satisfy the script witness requirement for some input locked by script hash H, you must either:
+ - Include a script that hashes to H in the `transaction_witness_set`, in fields 3, 6, or 7
+ - Include an input in field 18 of the transaction body (reference inputs), which refers to an input with `script_ref` set to a script that hashes to H
+
+Because of the use of the reference inputs, this will cause the construction of the plutus v1 script context to fail, as it has no backwards compatible way to expose these reference inputs to the script context.
+
+However, the `transaction_witness_set` is not exposed directly to the script context.
+
+So, we propose adding a new field for script references, as follows:
+
+```
+transaction_witness_set =
+  {
+  , ? 3: [* plutus_v1_script ]
+  ...
+  , ? 6: [* plutus_v2_script ]
+  , ? 7: [* plutus_v3_script ]
+  , ? 8: [* transaction_input ]
+  }
+```
+
+These inputs are not visible as reference inputs to the script context, and are *only* used to satisfy the script witness criteria. The node will look up each input referenced in `transaction_input`, and use any scripts found in the `script_ref` field, hash them, and use those scripts when it comes to evaluating whether each input can be spent.
+
+## Rationale: how does this CIP achieve its goals?
+
+This approach would immediately allow all major plutus v1 dApps to reduce their transaction sizes dramatically. Some napkin math for both Sundae and Minswap shows that this would cut around 85% of the transaction size for each transaction; Considering 90% of the space taken by by blocks is taken up by Plutus v1 scripts, this would have a massive impact on chain load.
+
+This may initially receive push-back because it admittedly has a "hacky" aesthetic; It feels like introducing multiple ways to accomplish the same thing, and adds a wrinkle to the specification of the transaction body.
```

Comment:
I think my proposed approach is less hacky, but my main concern with both approaches are that it opens us up to the risks discussed [here](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0031#how-should-we-present-the-information-to-scripts) (please reference that!), but overall it seems worth it.

That is, I think we can say "we have a general policy of not omitting information when constructing the script context, _except_ in a few very well-justified cases like this one". And that's still an okay place to be.